File: ArtificialIntelligence/hw1_notworking/GameModels/PegSolitaireRepresentation.py
#This class represents the game grid
#The grid is a list of lists. Pegs are represented as 1, void cells as 0.

class PegSolitaireRepresentation:
    centerX = 3  #static var. it is the column of the central space of play grid. needed for Manhattan Distance
    centerY = 3  #static var. it is the row of the central space of play grid. needed for Manhattan Distance
    
    def __init__(self, grid):
        self.grid = grid
        self.gridRows = len(grid)
        self.gridCols = len(grid[0])
        self.remainingBalls = self.remainingBalls()
        self.valf = -1
        
    def __hash__(self):
        l = ""
        for j in range(0,len(self.grid)):
            for k in range(0,len(self.grid[0])):
                if self.grid[j][k] != 2:
                    l += str(self.grid[j][k])
        h = int(l)
        return h
    
    def __repr__(self):
        rows = len(self.grid)
        cols = len(self.grid[0])
        t = ""
        for i in range(0,rows):
            toprint = ""
            for j in range(0,cols):
                toprint = toprint + " " + str(self.grid[i][j])
            t += toprint + "\n"
        return t
    
    def __eq__(self, other):
        for i in range(0, len(self.grid)):
            for j in range(0, len(self.grid[0])):
                if self.grid[i][j] != other.grid[i][j]:
                    return False
        return True
    
    def f(self, state):
        heuristic = state.getHeuristic()
        g = 1 #could be any number. this is the cost to move a peg
        h = heuristic.H(state)
        self.valf = g + h
        return self.valf
    
    def grid(self):
        return self.grid
    
    def pegExists(self, peg):
        val = self.grid[peg.getRow()][peg.getCol()]
        if val == 1:
            return True
        return False
    
    def remainingBalls(self):
        r = 0
        for i in range(0,len(self.grid)):
            for j in range(0, len(self.grid[0])):
                if self.grid[i][j] == 1:
                    r += 1
        return r
    
    
    

File: ArtificialIntelligence/hw1java/src/GameModel/PegSolitaireRepresentation.java
package GameModel;

import java.util.Arrays;

import Heuristic.Heuristic;

public class PegSolitaireRepresentation implements Representation {
	
	private int[][] grid;
	private int remainingBalls;
	private Heuristic heuristic;
	
	public PegSolitaireRepresentation(Heuristic heuristic, int[][] grid){
	    this.grid = grid;
	    this.heuristic = heuristic;
	    this.remainingBalls = this.getRemainingBalls();
	}

	@Override
    public boolean equals(Object other){
		PegSolitaireRepresentation otherrep = (PegSolitaireRepresentation) other;
        for(int i = 0; i < this.grid.length; i++)
            for(int j = 0; j < this.grid[0].length; j++) 
                if( this.grid[i][j] != otherrep.grid[i][j] )
                    return false;
        return true;
    }

    public int[][] grid(){
        return this.grid;
    }
    
    public boolean pegExists(Peg peg) {
        int v = this.grid[peg.getRow()][peg.getCol()];
        if( v == 1)
            return true;
        return false;
    }
    
    public int remainingBalls() {
    	return this.remainingBalls;
    }
    
    private int getRemainingBalls(){
        int r = 0;
        for (int i = 0; i < this.grid.length; i++)
            for ( int j = 0; j < this.grid[0].length; j++) 
                if( this.grid[i][j] == 1)
                    r += 1;
        return r;
    }

	public boolean isSolution() {
		if(this.remainingBalls == 1)
			return true;
		return false;
	}
	
	public String toString() {
		String r = "";
		for (int i = 0; i < this.grid.length; i++) {
            for ( int j = 0; j < this.grid[0].length; j++)
            	r += this.grid[i][j] + " ";
            r += "\n";
		}
		return r;
	}

	public int hashCode() {
		String r = "2";
		for (int i = 5; i < this.grid.length; i++)
            for ( int j = 0; j < this.grid[0].length; j++)
            	if(this.grid[i][j] != 2)
            		r += this.grid[i][j];
        int hashcode = Integer.parseInt(r);
		return hashcode;
	}
	
}
